### PR TITLE
Use Postgres LISTEN / NOTIFY

### PR DESCRIPTION
**Pull request by jgrund**
instead of long-polling / polling the database and performing many
table scans, we can use the LISTEN / NOTIFY mechanism of postgres
to be alerted to table changes.

We should start by replacing the health-route with a listen mechanism.

Note that this patch requires a version of the manager running:

https://github.com/whamcloud/integrated-manager-for-lustre/pull/130

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

### CODE REVIEW

File: db-utils.js [hunk 1]
```diff
@@ -8,14 +8,23 @@ const highland = require("highland");
 const broadcaster = require("./broadcaster");
 
 const pool = new Pool({ user: conf.DB_USER, password: conf.DB_PASSWORD, database: conf.DB_NAME, host: conf.DB_HOST });
+pool.on('error', e => {
+  console.error('Connection pool encountered an error', e.stack);
+  process.exit(-1);
```
Comment (jgrund) [inline db-utils.js]: should this be 1?

File: db-utils.js [hunk 2]
```diff
@@ -8,14 +8,23 @@ const highland = require("highland");
...
+});
+
 exports.pool = pool;
 
-const stream = highland((push, _) => {
+const stream = highland((push, _) => {
```
Comment (jgrund) [inline db-utils.js]: should this space be here?

File: socket-router/routes/health-route.js [hunk 3]
```diff
@@ -13,20 +13,13 @@ const highland = require("highland");
...
-      })
-      .then(x => {
-        return {
-          health: x.health,
-          count: x.num_alerts
-        };
-      })
-  );
+  return pool
+    .query("select * from health_status()")
```
Comment (jgrund) [inline socket-router/routes/health-route.js]: What happens if this query fails?

File: db-utils.js (hunk 1 above)
Comment (johnsonw) [reply db-utils.js]: Could be either I think. The docs show -1, but the more I think about it I'm starting to think we shouldn't have this. The docs state:
```
It is important you add an event listener to the pool to catch errors. Just like other event emitters, if a pool emits an error event and no listeners are added node will emit an uncaught error and potentially exit.
```
But with our approach I think that's exactly what we want to happen. Should I just remove it then?

File: db-utils.js (hunk 2 above)
Comment (johnsonw) [reply db-utils.js]: removed

File: socket-router/routes/health-route.js (hunk 3 above)
Comment (johnsonw) [reply socket-router/routes/health-route.js]: It will blow up and log the error.

File: db-utils.js (hunk 1 above)
Comment (jgrund) [reply db-utils.js]: what does potentially mean?

File: socket-router/routes/health-route.js (hunk 3 above)
Comment (johnsonw) [reply socket-router/routes/health-route.js]: The alternative is to add a `catch` to the bottom and log the error. I don't think we want that though do we?

File: db-utils.js (hunk 1 above)
Comment (johnsonw) [reply db-utils.js]: My interpretation of this statement is that if you don't handle the error your app will blow up.

File: socket-router/routes/health-route.js (hunk 3 above)
Comment (jgrund) [reply socket-router/routes/health-route.js]: do any clients leak if the program fails?

Comment (johnsonw) [reply socket-router/routes/health-route.js]: The docs say that this is the recommended way to make the query if a transaction is not taking place as you won't need to worry about leaking the client:
```
If you don't need a transaction or you just need to run a single query, the pool has a convenience method to run a query on any available client in the pool. This is the preferred way to query with node-postgres if you can as it removes the risk of leaking a client.
```
I'll take a quick look at the source and see what happens.

Comment (jgrund) [reply socket-router/routes/health-route.js]: right, but there is another client connection in db-utils.

Comment (johnsonw) [reply socket-router/routes/health-route.js]: Right, so let me see if there is a way to look at existing pools in postgres.

Comment (johnsonw) [reply socket-router/routes/health-route.js]: Ok, so I made an update where I'm listening for pool errors:
```
pool.on('error', e => {
  console.error('Connection pool encountered an error', e.stack);
  pool.end().then(() => {
    console.log('pool', pool);
    process.exit(-1);
  });
});
```
The output shows that all clients have been removed from the pool:
```
]: pool Pool {
Aug 17 22:24:47 adm.local node[27539]: domain: null,
Aug 17 22:24:47 adm.local node[27539]: _events: { error: [Function] },
Aug 17 22:24:47 adm.local node[27539]: _eventsCount: 1,
Aug 17 22:24:47 adm.local node[27539]: _maxListeners: undefined,
Aug 17 22:24:47 adm.local node[27539]: options:
Aug 17 22:24:47 adm.local node[27539]: { Client: { [Function: Client] super_: [Object], Query: [Object] },
Aug 17 22:24:47 adm.local node[27539]: user: 'chroma',
Aug 17 22:24:47 adm.local node[27539]: password: '',
Aug 17 22:24:47 adm.local node[27539]: database: 'chroma',
Aug 17 22:24:47 adm.local node[27539]: host: '/var/run/postgresql',
Aug 17 22:24:47 adm.local node[27539]: max: 10,
Aug 17 22:24:47 adm.local node[27539]: idleTimeoutMillis: 10000 },
Aug 17 22:24:47 adm.local node[27539]: log: [Function],
Aug 17 22:24:47 adm.local node[27539]: Client:
Aug 17 22:24:47 adm.local node[27539]: { [Function: Client]
Aug 17 22:24:47 adm.local node[27539]: super_:
Aug 17 22:24:47 adm.local node[27539]: { [Function: EventEmitter]
Aug 17 22:24:47 adm.local node[27539]: EventEmitter: [Circular],
Aug 17 22:24:47 adm.local node[27539]: usingDomains: false,
Aug 17 22:24:47 adm.local node[27539]: defaultMaxListeners: [Getter/Setter],
Aug 17 22:24:47 adm.local node[27539]: init: [Function],
Aug 17 22:24:47 adm.local node[27539]: listenerCount: [Function] },
Aug 17 22:24:47 adm.local node[27539]: Query: { [Function: Query] super_: [Object] } },
Aug 17 22:24:47 adm.local node[27539]: Promise: [Function: Promise],
Aug 17 22:24:47 adm.local node[27539]: _clients: [],
Aug 17 22:24:47 adm.local node[27539]: _idle: [],
Aug 17 22:24:47 adm.local node[27539]: _pendingQueue: [],
Aug 17 22:24:47 adm.local node[27539]: _endCallback: [Function: cb],
Aug 17 22:24:47 adm.local node[27539]: ending: true }
```

**Comment by johnsonw** [comment]
To test client connections to postgres I did the following:
1. Add four servers
2. Halt mds1 and wait for the alert to appear. journalctl showed the event taking place and the app reflected the change.
3. Start mds1 again and make sure the event change is picked up.
4. Query the database table to see the active client connections:
```
su - postgres -c "psql -c 'select * from pg_stat_activity;'"
 datid | datname  |  pid  | usesysid | usename  | application_name | client_addr | client_hostname | client_port |         backend_start         |          xact_start           |          query_start   
       |         state_change          | waiting |        state        |                                                                                                                                  
                                                                                                                                                                                                          
                                                                                                        query                                                                                             
                                                                                                                                                                                                          
                                                                                                                                              
-------+----------+-------+----------+----------+------------------+-------------+-----------------+-------------+-------------------------------+-------------------------------+------------------------
-------+-------------------------------+---------+---------------------+----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------
 16385 | chroma   |  2399 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:51:33.741291+00 |                               | 2018-08-20 17:51:33.800
162+00 | 2018-08-20 17:51:33.800306+00 | f       | idle                | LISTEN table_update
 16385 | chroma   |  8053 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:55:36.64546+00  |                               | 2018-08-20 17:55:36.663
879+00 | 2018-08-20 17:55:36.664718+00 | f       | idle                | COMMIT
 16385 | chroma   |  4920 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:58.391302+00 |                               | 2018-08-20 18:01:11.256
214+00 | 2018-08-20 18:01:11.271486+00 | f       | idle                | COMMIT
 12071 | postgres | 12752 |       10 | postgres | psql             |             |                 |          -1 | 2018-08-20 18:01:13.451831+00 | 2018-08-20 18:01:13.454091+00 | 2018-08-20 18:01:13.454
091+00 | 2018-08-20 18:01:13.454093+00 | f       | active              | select * from pg_stat_activity;
 16385 | chroma   |  4412 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:42.644773+00 |                               | 2018-08-20 18:01:12.337
912+00 | 2018-08-20 18:01:12.337987+00 | f       | idle                | COMMIT
 16385 | chroma   |  4411 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:42.643737+00 |                               | 2018-08-20 18:01:11.312
49+00  | 2018-08-20 18:01:11.312536+00 | f       | idle                | COMMIT
 16385 | chroma   | 25709 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:45:50.852246+00 | 2018-08-20 17:54:09.643158+00 | 2018-08-20 17:54:09.794
44+00  | 2018-08-20 17:54:09.797638+00 | f       | idle in transaction | SELECT "chroma_core_alertstate"."id", "chroma_core_alertstate"."record_type", "chroma_core_alertstate"."variant", "chroma_core_al
ertstate"."alert_item_type_id", "chroma_core_alertstate"."alert_item_id", "chroma_core_alertstate"."alert_type", "chroma_core_alertstate"."begin", "chroma_core_alertstate"."end", "chroma_core_alertstate
"."message", "chroma_core_alertstate"."active", "chroma_core_alertstate"."dismissed", "chroma_core_alertstate"."severity", "chroma_core_alertstate"."lustre_pid" FROM "chroma_core_alertstate" LEFT OUTER 
JOIN "chroma_core_alertemail_alerts" ON ("chroma_core_alertstate"."id" = "chroma_core_alertemail_alerts"."alertstate_id") LEFT OUTER JOIN "chroma_core_alertemail" ON ("chroma_core_alertemail_alerts"."al
ertemail_id" = "chroma_core_alertemail"."id") WHERE ("chroma_core_alertemail"."id" IS NULL AND "chroma_core_alertstate"."dismissed" = false )
 16385 | chroma   | 25716 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:45:50.940976+00 |                               | 2018-08-20 17:54:09.789
213+00 | 2018-08-20 17:54:09.797825+00 | f       | idle                | COMMIT
 16385 | chroma   |  4571 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:50.166354+00 |                               | 2018-08-20 18:01:12.343
859+00 | 2018-08-20 18:01:12.344381+00 | f       | idle                | COMMIT
 16385 | chroma   |  4574 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:50.266027+00 |                               | 2018-08-20 18:01:12.342
268+00 | 2018-08-20 18:01:12.342311+00 | f       | idle                | COMMIT
 16385 | chroma   |  4644 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:51.787193+00 |                               | 2018-08-20 17:57:48.221
424+00 | 2018-08-20 17:57:48.225884+00 | f       | idle                | COMMIT
(11 rows)
```
5. Stop the iml-realtime service
6. Check the active client connections again:
```
su - postgres -c "psql -c 'select * from pg_stat_activity;'"
 datid | datname  |  pid  | usesysid | usename  | application_name | client_addr | client_hostname | client_port |         backend_start         |          xact_start           |          query_start   
       |         state_change          | waiting |        state        |                                                                                                                                  
                                                                                                                                                                                                          
                                                                                                        query                                                                                             
                                                                                                                                                                                                          
                                                                                                                                              
-------+----------+-------+----------+----------+------------------+-------------+-----------------+-------------+-------------------------------+-------------------------------+------------------------
-------+-------------------------------+---------+---------------------+----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------
 12071 | postgres | 13105 |       10 | postgres | psql             |             |                 |          -1 | 2018-08-20 18:01:41.77009+00  | 2018-08-20 18:01:41.773905+00 | 2018-08-20 18:01:41.773
905+00 | 2018-08-20 18:01:41.773908+00 | f       | active              | select * from pg_stat_activity;
 16385 | chroma   |  8053 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:55:36.64546+00  |                               | 2018-08-20 17:55:36.663
879+00 | 2018-08-20 17:55:36.664718+00 | f       | idle                | COMMIT
 16385 | chroma   |  4920 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:58.391302+00 |                               | 2018-08-20 18:01:31.358
408+00 | 2018-08-20 18:01:31.359115+00 | f       | idle                | COMMIT
 16385 | chroma   |  4412 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:42.644773+00 |                               | 2018-08-20 18:01:39.889
062+00 | 2018-08-20 18:01:39.889141+00 | f       | idle                | COMMIT
 16385 | chroma   |  4411 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:42.643737+00 |                               | 2018-08-20 18:01:41.617
462+00 | 2018-08-20 18:01:41.617505+00 | f       | idle                | COMMIT
 16385 | chroma   | 25709 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:45:50.852246+00 | 2018-08-20 17:54:09.643158+00 | 2018-08-20 17:54:09.794
44+00  | 2018-08-20 17:54:09.797638+00 | f       | idle in transaction | SELECT "chroma_core_alertstate"."id", "chroma_core_alertstate"."record_type", "chroma_core_alertstate"."variant", "chroma_core_al
ertstate"."alert_item_type_id", "chroma_core_alertstate"."alert_item_id", "chroma_core_alertstate"."alert_type", "chroma_core_alertstate"."begin", "chroma_core_alertstate"."end", "chroma_core_alertstate
"."message", "chroma_core_alertstate"."active", "chroma_core_alertstate"."dismissed", "chroma_core_alertstate"."severity", "chroma_core_alertstate"."lustre_pid" FROM "chroma_core_alertstate" LEFT OUTER 
JOIN "chroma_core_alertemail_alerts" ON ("chroma_core_alertstate"."id" = "chroma_core_alertemail_alerts"."alertstate_id") LEFT OUTER JOIN "chroma_core_alertemail" ON ("chroma_core_alertemail_alerts"."al
ertemail_id" = "chroma_core_alertemail"."id") WHERE ("chroma_core_alertemail"."id" IS NULL AND "chroma_core_alertstate"."dismissed" = false )
 16385 | chroma   | 25716 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:45:50.940976+00 |                               | 2018-08-20 17:54:09.789
213+00 | 2018-08-20 17:54:09.797825+00 | f       | idle                | COMMIT
 16385 | chroma   |  4571 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:50.166354+00 |                               | 2018-08-20 18:01:39.890
469+00 | 2018-08-20 18:01:39.891035+00 | f       | idle                | COMMIT
 16385 | chroma   |  4574 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:50.266027+00 |                               | 2018-08-20 18:01:38.156
143+00 | 2018-08-20 18:01:38.156177+00 | f       | idle                | COMMIT
 16385 | chroma   |  4644 |    16384 | chroma   |                  |             |                 |          -1 | 2018-08-20 17:52:51.787193+00 |                               | 2018-08-20 17:57:48.221
424+00 | 2018-08-20 17:57:48.225884+00 | f       | idle                | COMMIT
(10 rows)
```

File: socket-router/routes/health-route.js [hunk 4]
```diff
@@ -3,66 +3,40 @@
...
-        return a;
-      }, [STATES.GOOD]);
-    }
-
-    function compare (a, b) {
-      var states = [STATES.GOOD, STATES.WARN, STATES.ERROR];
-
-      return states.indexOf(a) - states.indexOf(b);
-    }
+      .tap(x => console.error(x))
```
Comment (jgrund) [inline socket-router/routes/health-route.js]: This isn't needed.

File: socket-router/routes/host-route.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    stream
+      .tap(x => console.error(x))
+      .errors(pushSerializeError)
+      .through(through.unchangedFilter)
+      .each(x => resp.socket.emit(req.messageName, x));
+
+    next(req, resp, stream);
+  });
+};
```
Comment (jgrund) [inline socket-router/routes/host-route.js]: This looks like it shouldn't be in this PR.

File: db-utils.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    return c.query("LISTEN table_update").catch(e => console.error("got error", e));
+  });
+});
+
+exports.viewer = broadcaster(stream);
+
+const query = (query, transform) =>
+  pool
+    .query(query)
+    .then(r => r.rows[0])
```
Comment (jgrund) [inline db-utils.js]: Why are we hardcoding fetching the first row?

File: db-utils.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    .then(x => transform(x));
```
Comment (jgrund) [inline db-utils.js]: Why do we need to inline a transform fn?

**Pull request merged**